### DISCUSSION TITLE
Refactoring of the `capellambse.aird` and `capellambse.svg` modules

### DISCUSSION THREAD

**Discussion by Wuestengecko** [dbinfrago/py-capellambse#107] [Ideas]
This was already briefly touched on in #25, but we never opened a formal issue for this.

# Problem analysis

Currently, a lot of calculations are duplicated between `capellambse.aird` and `capellambse.svg`. To make matters worse, the two modules have small, but significant differences in their implementations. As an example, the `aird` module makes box sizing calculations without taking into account any icons, only the text label, whereas the `svg` module does account for icons. This has lead to several hard to debug rendering issues in the past already, and it's safe to assume that similar issues will continue to crop up in the future.

The fact that `aird` module has to do some calculations is unavoidable: The XML does not contain enough information on its own to render a complete diagram (which makes sense, as most of it can be relatively easily calculated from the information that is given). Examples include the physical extent of labels, i.e. their height and width when drawn – this can be (and is already) calculated based on the text, font and the font size.

On the other hand, the `svg` module needs to do its own additional calculations because not all of the information that it needs is provided to it from the `aird` module. For example, to draw text, it needs to not only know the size of the bounding box, but also the height and vertical position of each line. This effectively results in doing (almost) all the text-related calculations for the second time, and then doing some more on top.

# Proposed solution

A simple solution to this problem is for the `aird` module to provide all of the additional numbers that the `svg` module needs as part of the JSON document that is exchanged between the two. However, this approach makes the JSON format even more specific to the current SVG format converter, and does not scale very well if other modules emerge that provide different output formats (and which do not just convert from SVG, as is done by the current PNG converter).

A viable alternative approach is for the `svg` module to consume the `aird.Diagram` instance directly. This would allow us to provide either methods on that object or functions in some standardized place for all the calculations that need to be done additionally, while also taking advantage of the information already calculated before. Such methods can also be leveraged by alternative format conversion modules, without having to calculate things that are not needed for the particular requested format.

Both proposals constitute a breaking change for the JSON interface (the first one because older documents would be lacking crucial information, the second one because the interface would be abandoned entirely). However, I don't think that this is a particularly big issue, as the only real use case for these JSON documents was this exchange between `capellambse` submodules anyway. Without a way to reconstruct an `aird.Diagram` instance, they are not very useful for persisting modified (or newly created) diagrams, and with the `svg` module being the only known consumer of that format, it makes more sense to just convert it to the well-known, standardized SVG format straight away.

# Additional benefits

No matter which route we choose, addressing this issue can be expected to provide some additional benefits.

1. The current `svg` module is, frankly, a mess. During this refactoring, there is a great opportunity to clean it up, and to add some more documentation to it.
2. The `aird.parser` has grown to be a very complex beast as well. It's possible that some calculations can be moved out of it, reducing its overall complexity.
3. By removing the need to perform the same calculations twice, we should experience a noticeable speed-up in the SVG rendering process.

**Comment by vik378** [comment]
would getting rid of JSON interface have any negative effects on context-diagrams extension?  if I got it right we used the same interface spec to render the outputs of elk or did I get it wronk, @ewuerger ?

**Reply by ewuerger** [reply]
Absolutely not. With the context-diagram extension we are building a special JSON solely for ELK. Afterwards an `aird.Diagram` is build as the final stage. Then we rely on the render functionality implemented by capellambse.

In capellambse we decode the `aird.Diagram` again as JSON which the svg submodule accepts. I agree with Martin that this is really unnecessary. Cildhood desease that we should cure.

The JSON format as an export feature can still be maintained/kept.

**Reply by Wuestengecko** [reply]
> The JSON format as an export feature can still be maintained/kept.

I wonder how much sense that actually makes. Just keeping it for the sake of keeping it is IMO a bad idea, because it would just be a maintenance burden for us without any real use case. But if we were to extend and modify the format to allow to fully reconstruct an `aird.Diagram` from it, then I can see it becoming useful as a storage or exchange format. We're far away from that though, and I'd argue that the changes to the format would be so fundamental that it'd make more sense to start from scratch and design a new one with the specific requirements in mind.

**Reply by vik378** [reply]
agree, no long term need for that diagram format. Lets do it then
